Pipfile semantics (2.0 Rewrite)

masenf opened the pull request:
# The rewrite that no one asked for? 🤷 

  * Plugin no longer using `pipenv` to create virtual environments
  * Plugin no longer using `pipenv` to run commands
  * Only activate `pipenv` when `Pipfile`, `Pipfile_{envname}` or `Pipfile_{envname}.lock` is present in `toxinidir`
  * Do NOT activate `pipenv` when test environment specifies `deps`
  * Automatically `pipenv sync` (from `Pipfile_{envname}.lock`, if present)
    * Enable strict-pinning workflow with multiple python versions
    * New tox CLI option: `--pipenv-update` to explicitly (re)lock dependencies for multiple environments
  * testenv ini-setting, `skip_pipenv`: opt-out of all `pipenv` operations for the environment
  * Customize `pipenv` `install_deps` command:
    * testenv ini-setting `pipenv_install_opts` / `TOX_PIPENV_INSTALL_OPTS`
    * Environment variables take precedence over testenv ini-setting
  * Code restructuring for better modularity and readability

# Testing

* 100% branch coverage with unit tests
* 98% branch coverage with integration tests

```
py37 run-test: commands[0] | pytest -x test/integration --cov
====================================================================== test session starts ======================================================================
platform darwin -- Python 3.7.15, pytest-7.2.0, pluggy-1.0.0
cachedir: .tox/py37/.pytest_cache
Using --randomly-seed=48219505
rootdir: /Users/masen/code/tox-dev/tox-pipenv, configfile: tox.ini
plugins: mock-3.10.0, xdist-3.0.2, randomly-3.12.0, cov-4.0.0
collected 58 items                                                                                                                                              

test/integration/test_integration.py ..........................................................[100%]
---------- coverage: platform darwin, python 3.7.15-final-0 ----------
Name                     Stmts   Miss Branch BrPart  Cover   Missing
--------------------------------------------------------------------
tox_pipenv/__init__.py       1      0      0      0   100%
tox_pipenv/plugin.py       135      1     50      2    98%   142->144, 145
--------------------------------------------------------------------
TOTAL                      136      1     50      2    98%
================ 58 passed in 1071.11s (0:17:51) ==============
```

masenf committed:
conftest: remove unused private import

pytest 5+ compatibility

masenf committed:
conftest: use a separate toxinidir and envdir

allows for writing tests that _clone_pipfile from toxinidir to the
envdir (otherwise .copy() says they're the same file...)

masenf committed:
conftest: track action setactivity state

allow tests to assert on parameters passed to setactivity

masenf committed:
test_testenv_install_deps: mocker.patch.dict("os.environ")

avoid breaking `os.environ` dict-semantics to allow environment
variables to be used for configuration values

masenf committed:
`pipenv sync` when Pipfile.lock exists

check for the lockfile in
  * {toxinidir}/Pipfile.lock.{envname}
  * {toxinidir}/Pipfile.lock

if any of these files exist, copy them into the `{envdir}` and use
`pipenv sync` to install a virtualenv that matches the lock file.

masenf committed:
version 1.11.0a1

masenf committed:
TOX_PIPENV_INSTALL_ARGS are used instead of `pipenv install --dev`

If the caller wants to customize the `pipenv install` behavior, the
plugin will now respect the environment variable
TOX_PIPENV_INSTALL_ARGS.

If this variable is present, it is used during the
`tox_testenv_install_deps` hook to install dependencies into the test
environment via `pipenv`.

Note: if the args do not contain "sync", then the dependency list
will be appended to the given command.

masenf committed:
remove testenv_create and runtest hookimpl

These aren't useful if we allow tox to manage the venv and just use
pipenv to install and lock packages within it.

masenf committed:
version 1.11.0a2

masenf committed:
set PIPENV_VERBOSITY=-1

ignore the "you're running in a virtualenv" message

masenf committed:
Show path to Pipfile.lock in installdeps action

display extra information for the user to track down any installation
issues that may arise.

masenf committed:
version 1.11.0a3

masenf committed:
tox-pipenv v2 rewrite: --pipenv-lock and limit scope

* only run when Pipfile or Pipfile.lock.{envname} is present
* explicitly lock dependencies when `--pipenv-lock` is given
  * copy environment-specific lock file back to toxinidir
* testenv ini-settings
  * skip_pipenv
  * pipenv_install_cmd
  * pipenv_install_args
* overriding the install command is now split between command and args,
  environment variables take precedence over testenv setting
* restructuring for better modularity and readability

masenf committed:
Rewrite README for tox-pipenv v2

Move markdown for easier maintenance

----------------------

HISTORY: add info on 2.0.0 rewrite

convert to markdown for easier maintenance

update long_description mime type and extensions in setup.py

masenf committed:
Respect PIPENV_PIPFILE

if the caller provides PIPENV_PIPFILE, then use that path and base name
instead of looking in toxinidir and assuming the name "Pipfile".

update pipenv_install_args to pipenv_install_opts throughout

masenf committed:
Modernize packaging PEP 621

Primarily use `pyproject.toml` for package metadata.

Retain slim downed `setup.py` for editable install
support.

masenf committed:
pyproject.toml: publish as tox-pipenv2

for testing the rewrite branch until it is accepted upstream

masenf committed:
plugin: cleanup unused imports

masenf committed:
README.md: fixup pipenv_install_cmd type

Move Masen Furer to the top the Authors list,
as the author of the rewrite (2.0) branch.

masenf committed:
tox.ini: use pytest-cov

remove `coverage` environment

move tool configuration from `setup.cfg` to `tox.ini`

masenf committed:
conftest: parametrize venv fixture

fixtures `has_pipfile` and `has_pipfile_lock` are
parametrized and modify `venv` to create a test
matrix of option combinations.

move all mock object properties to instance variables
to avoid leaking data across test cases

masenf committed:
plugin: DEFAULT_PIPENV_INSTALL_OPTS is now tuple

avoid accidentally mutating the default opts

apply `--pre` to either `install` or `sync` (but
don't mutate the default)

masenf committed:
plugin: init global PIPFILE on import

even if it's not initialized _exactly_ correct,
ensure that the values are usable, even before
tox_configure hook is called

masenf committed:
test_testenv_install_deps: rewrite unit tests

cover basic cases with mock venv and Popen

masenf committed:
test_testenv_install_deps: extended test cases

tests for pipenv_lock and skip_pipenv

masenf committed:
test cases for configure, _pipenv_env, and runenvreport

masenf committed:
plugin: raise ToxPipenvError when Pipfile not available

for operations that require a Pipfile, raise an
exception when one is not specified

split up _init_global_prefix for testability.

masenf committed:
tox.ini: avoid test warnings

* pass --strict-markers
* fail_under (not fail_user)

masenf committed:
lint: update lint environment and blacken

masenf committed:
tox: docs and publish environments

repeatable doc and package builds via tox

masenf committed:
XXX: README / __init__.py tox-pipenv2 branding

masenf committed:
plugin: _should_skip returns a reason now

raise an exception if the environment will skip
pipenv, but pipenv-influencing options are specified

masenf committed:
tox_pipenv: newline at end of __init__.py

black formatting fixup

masenf committed:
tox.ini: use `source_pkgs` in [coverage:run] section

Since the package name "tox_pipenv" and the
source directory are the same, passing
`--cov tox_pipenv` results in pytest-cov taking
the arg as a filesystem path, rather than a package
name. Using a filesystem path here is inconvenient
because the package gets installed into a virtualenv
and doesn't run from the repo root.

The previous workaround was to explicitly pass
the {envsitepackagesdir} to `--cov`, but _this_
technique didn't work for subprocesses that
change the working directory (see pytest-cov#465).

This workaround resolves the subprocess coverage
issue, but retains the paths mapping so that the
output references the source files in the top-level
tox_pipenv directory.

masenf committed:
test_integration: end-to-end tests using `tox`

Test various combinations of Pipfile present,
Pipfile.lock present, and --pipenv-lock specified.

tucked commented on pyproject.toml:
```diff
@@ -2,5 +2,47 @@
+build-backend = 'setuptools.build_meta'
+
+[project]
+name = "tox-pipenv2"
```
Not personally a fan of the major-version-in-dist-name pattern FWIW

masenf replied:
this is so i can publish on PyPI and test until the code is accepted into tox-dev/tox-pipenv

i'm hoping the upstream will take the rewrite and allow the use of the original repo and `tox-pipenv` name 🤞

masenf committed:
test_integration: end-to-end tests with deps =

these tests should always skip pipenv and sometimes
raise exceptions if pipenv actions are requested
but pipenv wont be used for whatever reason.

masenf committed:
test_integration: install_opts and install_cmd

test that overriding the install command or
options via environment variable is reflected
in the underlying command

masenf committed:
tox.ini: use pytest-xdist

attempt to run in less time...

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+## `{toxinidir}/Pipfile`
+
+Use `pipenv` to create a `Pipfile` in the project directory, next to `tox.ini`.
```
Should a note be added about Python version requirements?
``` console
$ pipenv install requests
Creating a virtualenv for this project...
Pipfile: /tmp/tmp.WgS1agFV7j/Pipfile
Using /ifs/home/dtucker/.pyenv/versions/3.10.0/bin/python (3.10.0) to create virtualenv...
...
Creating a Pipfile for this project...
...
$ tail -n2 Pipfile
[requires]
python_version = "3.8"
```
I think that should be respected / manually removed for multi-Python.

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+### Yes
+
+In the absense of `Pipfile` in `toxinidir`, this plugin won't do anything!
```
I wonder if it would make sense (not necessarily right now) to `pipenv install` anything in `deps` to create a (temporary?) `Pipfile`. I don't think it'd be a problem even if a `Pipfile` already exists:
``` console
$ cp Pipfile{,.old}
$ pipenv install requests
Installing requests...
Adding requests to Pipfile's [packages]...
✔ Installation Succeeded
Installing dependencies from Pipfile.lock (5cd377)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
$ diff -s Pipfile{,.old}
Files Pipfile and Pipfile.old are identical
```

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Note: `Pipfile.lock` is _ignored_
+
+It doesn't make sense to share a lock file between environments, so an
```
FWIW, we can tell what version `Pipfile.lock` is associated with 🤔 
``` json
{
    "_meta": {
        "hash": {
            "sha256": "95bd67d0717652fcbc747c8dbdc1250b249e4d6774919125caefa1056b5cd377"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "3.8"
        },
        "sources": [
...
```

masenf committed:
unit/test_addoption: smoke test for coverage

ensure that `unit` tests provide 100% coverage of
the plugin code

masenf committed:
rename `--pipenv-lock` to `--pipenv-update`

instead of separately calling `pipenv lock` and
then using the `sync` behavior based on the presence
of the new lock file, just use `update`, since that's
what it was made for.

this also simplifies the implementation and the
tests, since there is no longer an option that might
run a 2nd command

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+If true, this plugin will not take any action in the environment.
+
+## `pipenv_install_cmd`
+
+_string_. Specified per `[testenv]` section.
+
+Override the sub-command passed to `pipenv` during the `install_deps` stage.
+
+By default, the plugin will use `sync` when a lockfile is present and `install`
+otherwise.
+
+### `TOX_PIPENV_INSTALL_CMD`
+
+_string_. Environment Variable.
+
+Global override for `pipenv_install_cmd` will apply to all environments.
+
+## `pipenv_install_opts`
+
+_argv_. Specified per `[testenv]` section.
+
+Override the args passed to `pipenv` during the `install_deps` stage.
+
+By default, the plugin will use `--dev` to install all dependencies specified
+in the `Pipfile`.
+
+### `TOX_PIPENV_INSTALL_OPTS`
+
+_string_. Environment Variable.
+
+Global override for `pipenv_install_opts` will apply to all environments.
```
Should these be combined? The example in my head that causes discord here is `install --deploy` (there is no `sync --deploy`).

Without this plugin, if the user edits `deps`, `tox` figures out that it needs to `recreate` automatically.
With this plugin, if the user updates `Pipfile` or `Pipfile.lock`, nothing happens without manual intervention (e.g. `--recreate` or `--pipfile-lock`), and even with `-r`, there's no guarantee that the `Pipfile.lock` matches the `Pipfile`.

masenf replied:
this is a good point. I have removed the `TOX_PIPENV_INSTALL_CMD` and `pipenv_install_cmd` settings.

The semantics of `pipenv_install_opts` have changed:

If this setting is provided, it overrides _whatever_ the plugin was going to do with `pipenv` for `install_deps`. So you would supply it as `TOX_PIPENV_INSTALL_OPTS="install --deploy"` and expect it to do the right thing.

The only wrinkle remaining here is the `pip_pre` option in the envconfig. With the change described above, I'm NOT adding `--pre` if the caller overrides the `pipenv_install_opts`, but I don't _know_ if that's the right thing. Particularly in the `install --deploy` case, it seems like the envconfig specifying `pip_pre` would be important. However if the caller provides some _other_ set of opts that is incompatible with `--pre` then it makes sense (can't think of scenario at the moment)

tucked commented on README.md:
````diff
@@ -0,0 +1,244 @@
+updating dependency pins.
+
+```
+tox --recreate --pipenv-lock
````
It'd be nice to be able to manage the lock file with `pipenv` natively and still have things work (see my note about `--deploy`).

masenf replied:
`--pipenv-lock` is now `--pipenv-update` and works like `pipenv update`, except instead of updating `Pipfile.lock`, it updates `Pipfile.lock.{envname}`.

(`--pipenv-lock` is removed entirely)

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+Override the sub-command passed to `pipenv` during the `install_deps` stage.
+
+By default, the plugin will use `sync` when a lockfile is present and `install`
```
Is there a reason to use `sync` instead of just `install` with no arguments? I think being able to always use `install` would be super clean.

masenf replied:
new diff will use `install --ignore-pipfile` instead of `sync`.

basically the idea is, if you have an environment-specific lock file, you definitely want to install from the lock file, _unless_

a) you specify `--pipenv-update` to update the lock file
b) you specify `TOX_PIPENV_INSTALL_OPTS="install --deploy"`, to check if the lock file is up-to-date

(there's a possibility that b) becomes `--pipenv-deploy`, but not sure if we need it)

tucked commented on README.md:
```diff
@@ -0,0 +1,244 @@
+
+Override the args passed to `pipenv` during the `install_deps` stage.
+
+By default, the plugin will use `--dev` to install all dependencies specified
```
IMO, a major version bump would be an excellent opportunity to NOT carry this forward... Why should the tox default be different than the `pipenv` one?

Also/Alternatively, would it be weird to co-opt [`usedevelop`](https://tox.wiki/en/latest/config.html#conf-usedevelop) and/or [`--develop`](https://tox.wiki/en/latest/config.html#cmdoption-tox-develop) for this?

masenf replied:
nixed, need to update the README

tucked reviewed (commented):
Very cool stuff 😎

masenf committed:
remove unused integration test files

masenf committed:
test_integration: additional end-to-end coverage

* skip_pipenv
* pip_pre in tox.ini
* PIPENV_PIPFILE set to alternate name
* no deps given, no Pipfile available

masenf committed:
plugin: replace "sync" with "install --ignore-pipfile"

unless the user requests update, then the plugin
will use "install" regardless of the presence of
the lock file;

avoid asymmetry w.r.t pipenv_install_args, which
might be fed to a different command depending on
lock file state

masenf committed:
plugin: remove pipenv_install_cmd

for sanity, pipenv_install_opts will take over the
entire _install_args command line and no
additional opts will be added by the plugin,
this is an escape hatch.

this allows a caller to pass

TOX_PIPENV_INSTALL_OPTS="install --deploy"

and know that the plugin wasn't going to use "update"
or some other root command, based on the repo state

plugin: raise exception when --pipenv-update and pipenv_install_opts are combined

These options can't be reasonably combined unless
"update" is in the opts.

masenf committed:
README: remove pipenv_install_cmd description

this setting was removed entirely

masenf committed:
HISTORY: update with recent tweaks

masenf committed:
integration: split up data and fixtures

masenf committed:
unit/conftest: rewrite mocks with attrs

masenf committed:
.github: modernize workflows

add publish to pypi workflow

masenf committed:
.github: modernize workflows (fixup)

add publish to pypi workflow

masenf committed:
test_addoptions: py3.7 compatible mock assertions

masenf committed:
Support env-specific Pipfile

revamp support for environment-specific lock file in the process:

Instead of appending `.{envname}` to the lock file, append `_{envname}`
to the Pipfile (and fallback to `Pipfile` if the env-specific one
doesn't exist).

Then the lock file naturally becomes Pipfile_{envname}.lock, which can
still be used with native `pipenv` by setting
`PIPENV_PIPFILE=Pipfile.lint pipenv update` (for example)

masenf commented:
> Plugin no longer using pipenv to create virtual environments

Thinking about some more use cases over the weekend and this removal will have to be reverted.

Creating the virtualenv using pipenv is literally the one thing a plug-in can do, that you can't easily do directly in tox.ini.

I'm also realizing that "pipenv_install_opts" can likely be handled by the built-in install_command setting (and this could enable the 'initialize Pipfile from deps' workflow without additional code). This probably also simplifies the tests a lot, so will try this one before bringing back tox_testenv_create.

masenf commented:
removal of `pipenv_install_opts` in https://github.com/masenf/tox-pipenv/pull/3
